docs/architecture-decision-records/0003-generalize-modules.md

```diff
@@ -65,23 +68,22 @@ This approach:
         // Note, however, that the implementation may produce output events even without receiving any input.
         // Before Run is called, ActiveModule guarantees not to read or write,
         // respectively, from eventsIn and to eventsOut.
+        //
+        // In general, Run is expected to block until explicitly asked to stop
+	    // (through a module-specific event or by canceling ctx)
+        // and thus should most likely be run in its own goroutine.
         // Run must return after ctx is canceled.
-        Run(ctx context.Context, eventsIn <-chan *events.EventList, eventsOut chan<- *events.EventList) error
-
-        // BackOff asks the module to back off, i.e., to not produce new events
-        // except in immediate reaction to input events, until ctx is canceled.
-        // After BackOff is called and before ctx is canceled,
-        // the module will only produce a bounded number of output events and only in reaction to processing input events.
-        // In other words, if BackOff has been called, ctx has not been canceled, and the input channel is not written to,
-        // the module will eventually stop producing output events.
         //
-        // This function is necessary for guaranteeing that the internal Node event buffers do not grow indefinitely.
-        // This could potentially happen when external events are injected into the system through active modules
-        // faster than the system can process them.
-        // In such a case, the Node asks all active modules to back off
-        // until enough events in the internal buffers have been processed.
-        BackOff(ctx context.Context)
-        
+        // Note that the node does not guarantee to always read events from eventsOut.
+        // The node might decide at any moment to stop reading from eventsOut for an arbitrary amount of time
+	    // (e.g. if the Node's internal event buffers become full and the Node needs to wait until they free up).
+        //
+        // The implementation of Run is required to always read from eventsIn.
+        // I.e., when an event is written to eventsIn, Run must eventually read it, regardless of the module's state
+        // or any external factors
+        // (e.g., the processing of events must not depend on some other goroutine reading from eventsOut).
```

Comment:
What if the module may require significant time to fully process incoming events (e.g. WAL or application) and other modules keep producing new events for that module?

Reply:
If that module's buffer becomes too big, external events will be put on hold (by stopping to read from active modules' output channels), eventually leading to the other modules also stopping to produce new events (since they always only react to inputs). If modules are properly implemented, this is guaranteed to happen. If they end up "pumping" each other (producing ever-increasing numbers of events in reaction to each other), that's a different issue and definitely a bug in module implementation.

Reply:
Ah, that was what you meant by "eventually read".